Fix unpack_defs handling of nested $ref inside anyOf items

colesmcintosh opened the pull request:
## Title

Fix unpack_defs handling of $ref inside anyOf/items structures

## Relevant issues

Fixes #11372

## Pre-Submission checklist

- [x] I have added testing in the `tests/litellm/` directory (see `test_unpack_defs_resolves_nested_ref_inside_anyof_items`)
- [x] I have added a screenshot of my new test passing locally (placeholder below)
- [x] My PR passes all unit tests on `make test-unit`
- [x] My PR's scope is isolated to solving this specific bug

<img width="909" alt="Screenshot 2025-06-22 at 7 42 14 AM" src="https://github.com/user-attachments/assets/5c77b0bb-7808-4dc0-8295-15c7bbd8cfa6" />

## Type

🐛 Bug Fix

## Changes

1. Updated `unpack_defs` in `litellm/litellm_core_utils/prompt_templates/common_utils.py` to recursively resolve `$ref` references appearing inside `items` within `anyOf` branches and removed early return that skipped non-property schemas.
2. Added concise inline comments; removed large comment blocks.
3. Added unit test `test_unpack_defs_resolves_nested_ref_inside_anyof_items` in `tests/test_litellm/test_utils.py` to cover the regression.

---

This fix ensures complex Pydantic schemas with optional lists (e.g., `Optional[List[VatAmount]]`) are correctly expanded before being sent to providers like VertexAI, preventing empty `items` objects and maintaining schema integrity.

colesmcintosh committed:
refactor(unpack_defs): enhance handling of schema properties and anyOf structures

- Improved the unpack_defs function to handle top-level properties and nested structures more effectively.
- Added recursion for items in schemas and refined the handling of anyOf branches to ensure proper unpacking of references.
- Streamlined the logic for resolving $ref keys and managing nested schemas.

colesmcintosh committed:
test(unpack_defs): add test for resolving nested $ref in anyOf items

- Introduced a new test to verify that unpack_defs correctly resolves references within items of anyOf structures, addressing a specific bug scenario (Issue #11372).
- The test includes a minimal schema to ensure proper unpacking and validation of the resolved items schema.

zseymour commented:
Just wanted to call out #11958 is also open and addresses this problem from a different angle, mainly just resolving all references in the JSON schema, regardless of if the key is called `$defs` (some tools are still using `definitions` for example) and supporting additional clauses (like `allOf` in addition to `items` and `anyOf`)

colesmcintosh committed:
Merge branch 'BerriAI:main' into fix-11372-unpack-defs-ref-in-anyof

colesmcintosh committed:
refactor(unpack_defs): implement a generic resolver for $ref entries

- Redesigned the unpack_defs function to provide a more robust and dependency-free implementation for resolving all $ref entries in JSON schemas.
- Introduced a depth-first traversal method that efficiently handles nested structures, including anyOf, allOf, and items, while avoiding infinite recursion.
- Enhanced memory management by resolving nodes in-place without creating a full dereferenced copy, improving performance and reducing overhead.

colesmcintosh commented:
@zseymour Thanks for flagging [#11958](https://github.com/BerriAI/litellm/pull/11958). I’ve incorporated a similar `$ref` resolution into this PR with a much lighter touch:

* Introduced a single depth-first walker (no new deps) that expands any `$ref` regardless of where it appears (`items`, `anyOf`, `allOf`, `oneOf`, `additionalProperties`, etc.).  
* Supports both `$defs` and`definitions`.  
* Cycle-safe via a `seen` set.  
* In-place mutation, so no call-site changes needed.  

If there are other schema shapes you’d like covered, let me know and I can add parametric tests. Otherwise, I believe this is the lighter, safer fix for #11372.

colesmcintosh commented:
The new depth-first walker expands every `$ref` (`items`, `anyOf`, `allOf`, etc.). The single regression test here is enough; extra fixtures would just slow CI. Unless there’s a missed edge-case, let's keep tests as-is and close #11958.

krrishdholakia commented on tests/test_litellm/test_utils.py:
wrong place for test - it should be in here - https://github.com/BerriAI/litellm/blob/main/tests/test_litellm/litellm_core_utils/prompt_templates/test_litellm_core_utils_prompt_templates_factory.py

colesmcintosh committed:
Merge branch 'BerriAI:main' into fix-11372-unpack-defs-ref-in-anyof

colesmcintosh committed:
Remove test for unpack_defs resolving nested references in anyOf items from test_utils.py

colesmcintosh committed:
Add test for unpack_defs resolving nested references in anyOf items

This commit introduces a new test to ensure that the unpack_defs function correctly resolves $ref references within items of anyOf schemas, addressing issue #11372. The test verifies that the unpacked schema contains the expected properties and structure.
